In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv('/content/netflix_titles.csv')

In [3]:
df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

##📊 Netflix Movies & TV Shows Dataset Summary

- Total Entries (Rows): ~8,807

- Columns: 12
(show_id, type, title, director, cast, country, date_added, release_year, rating, duration, listed_in, description)

###🔎 Column-wise Summary

- show_id

Unique ID har show/movie ka (no missing, unique).

- type

Categories: Movie, TV Show.

- Movies > TV Shows (Movies zyada hain).

- title

- Name of movie/show.

Mostly unique, kuch duplicate ho sakte hain (different countries ke version).

- director

Bohat missing values (many movies/shows without director info).

- cast

Missing data hai (kuch shows me actors ka naam nahi diya).

Top actors multiple baar repeat hote hain.

- country

Missing values present.

Dominated by United States, India, UK, South Korea.

- date_added

Netflix pe add hone ki date.

Format inconsistent (strings, convert to datetime).

- release_year

Show/movie ka original release year.

- Range: ~1925 – 2021.

- rating

- Audience rating (e.g., TV-MA, PG-13, R).

Missing values present, thodi inconsistent naming bhi hai.

- duration

Movies → Minutes (e.g., 90 min).

TV Shows → Seasons (e.g., 2 Seasons).

- listed_in

Genres/categories (multi-label field, e.g., Comedies, Action, Thrillers).

Messy, multiple genres ek cell me.

- description

###1). Short summary of the show/movie.

Mostly clean text.

⚠️ Issues (Uncleaned Data)

Missing values: director, cast, country, rating.

Inconsistent data: date_added as string, duration format differs (min vs seasons).

Multi-valued columns: listed_in has multiple genres in one cell.

Duplicates possible: Some title entries.

###2). ✅ Insights You Can Extract

Count of Movies vs TV Shows.

Top countries producing Netflix content.

Trend of content added over years (date_added).

Most common genres.

Distribution of ratings (PG, R, TV-MA, etc.).

Longest movies / highest seasons shows.

## a). MESSAY DATA

- 📌 Structural Issues in Netflix Dataset

- Duration column (mixed values)

- Movies → minutes (e.g., 90 min)

- TV Shows → seasons (e.g., 2 Seasons)

- Problem → ek column me do alag types ki information hai → ideally split into two: duration_minutes & num_seasons.

- Listed_in column (genres/categories)

- Multiple categories ek hi cell me comma-separated (e.g., Comedies, Dramas, Thrillers).

- Problem → 1 cell ≠ 1 value (violates database normalization).

Ideally split into multiple rows or separate genre table.

- Cast column

Multiple actors ek hi cell me (e.g., "Shah Rukh Khan, Kajol, Rani Mukerji").

Problem → many-to-many relationship should not be in one column.

- Country column

Sometimes contains multiple countries (e.g., "United States, India").

Problem → should be split into separate rows for each country.

- Date_added column

Stored as text (e.g., "September 24, 2021").

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


##📌 Netflix Dataset Summary

###1).🔎 Missing Values (per column)

- show_id → 0 nulls ✅

- type → 0 nulls ✅

- title → 0 nulls ✅

- director → 2634 nulls ⚠️

- cast → 825 nulls ⚠️

- country → 831 nulls ⚠️

- date_added → 10 nulls ⚠️

- release_year → 0 nulls ✅

- rating → 4 nulls ⚠️

- duration → 3 nulls ⚠️

- listed_in → 0 nulls ✅

- description → 0 nulls ✅

###2).🔎 Datatype Issues

- show_id → object → ✅ correct (unique string IDs)

- type → object → ✅ correct (categorical: Movie/TV Show)

- title → object → ✅ correct (string names)

- director → object → ✅ correct (string names)

- cast → object → ✅ correct (string names)

- country → object → ✅ correct (string names)

- date_added → object ❌ → should be datetime64[ns]

- release_year → int64 → ✅ correct (numeric year)

- rating → object → ✅ correct (categorical)

- duration → object ❌ → should be split into:

- duration_minutes (int64) for Movies

- num_seasons (int64) for TV Shows

- listed_in → object → ✅ correct (string categories, - later split if needed)

- description → object → ✅ correct (text summary)

In [5]:
# ---------- 1. date_added ----------
# strip spaces + convert to datetime safely
df['date_added'] = pd.to_datetime(df['date_added'].str.strip(), errors='coerce')

# jo NaT (null) bache unko release_year se fill karo
df['date_added'] = df.apply(
    lambda row: pd.to_datetime(f"01-01-{row['release_year']}")
                if pd.isnull(row['date_added']) else row['date_added'],
    axis=1
)

# ---------- 2. director ----------
# null ko "Unknown Director" se fill
df['director'] = df['director'].fillna("Unknown Director")

# ---------- 3. cast ----------
# null ko "No Cast Info" se fill
df['cast'] = df['cast'].fillna("No Cast Info")

# ---------- 4. country ----------
# null ko "Unknown Country" se fill
df['country'] = df['country'].fillna("Unknown Country")

# ---------- 5. rating ----------
# forward fill lagao (pichle row ka value uthao)
df['rating'] = df['rating'].fillna(method='ffill')

# ---------- 6. duration ----------
# null ko "Unknown Duration" se fill
df['duration'] = df['duration'].fillna("Unknown Duration")

/tmp/ipython-input-3484330129.py:26: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df['rating'] = df['rating'].fillna(method='ffill')


In [6]:
#Test:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   object        
 1   type          8807 non-null   object        
 2   title         8807 non-null   object        
 3   director      8807 non-null   object        
 4   cast          8807 non-null   object        
 5   country       8807 non-null   object        
 6   date_added    8807 non-null   datetime64[ns]
 7   release_year  8807 non-null   int64         
 8   rating        8807 non-null   object        
 9   duration      8807 non-null   object        
 10  listed_in     8807 non-null   object        
 11  description   8807 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(10)
memory usage: 825.8+ KB


In [7]:
# 1. date_added → datetime format me convert
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')

# 2. duration → split into two columns
df[['duration_int', 'duration_type']] = df['duration'].str.extract(r'(\d+)\s*(\w+)')

# duration_int ko numeric me convert karna (int)
df['duration_int'] = pd.to_numeric(df['duration_int'], errors='coerce')

In [8]:
df.describe()

date_added  release_year  duration_int
count                           8807   8807.000000   8804.000000
mean   2019-05-13 21:37:54.781423872   2014.180198     69.846888
min              2003-01-01 00:00:00   1925.000000      1.000000
25%              2018-04-03 00:00:00   2013.000000      2.000000
50%              2019-07-01 00:00:00   2017.000000     88.000000
75%              2020-08-18 00:00:00   2019.000000    106.000000
max              2021-09-25 00:00:00   2021.000000    312.000000
std                              NaN      8.819312     50.814828

In [9]:
#DRop value for missing:
df = df.dropna(subset=['duration_int', 'duration_type'])


In [10]:
#Test:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8804 entries, 0 to 8806
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   show_id        8804 non-null   object        
 1   type           8804 non-null   object        
 2   title          8804 non-null   object        
 3   director       8804 non-null   object        
 4   cast           8804 non-null   object        
 5   country        8804 non-null   object        
 6   date_added     8804 non-null   datetime64[ns]
 7   release_year   8804 non-null   int64         
 8   rating         8804 non-null   object        
 9   duration       8804 non-null   object        
 10  listed_in      8804 non-null   object        
 11  description    8804 non-null   object        
 12  duration_int   8804 non-null   float64       
 13  duration_type  8804 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(11)
memory usage: 1.0+ MB


In [26]:
df.to_csv('cleaned_netflix_data.csv', index=False)